feat: Use component-based effect for 'boost' effects

skaldarnar opened the pull request:
Extracted from #14
Based on #15

Replace the copy-paste implementation of the alteration effect giving a health boost with the common component-based base class.

:exclamation: This moves some code around and thus changes the logic. Needs careful review (although I believe that the previous implementation was erroneous).

skaldarnar committed:
feat: Introduce a ComponentBasedAlterationEffect

Extracted from #14

This is a common base class for alteration effects based on a single component. It only requires implementations of component _upsert_ and _update_ to fully define an effect (in addition to the component class and the general effect id). _Removing_ the component can also be customized.

The implementation is extracted from WalkSpeedAlterationEffect and can be found in other effect classes such as SwimSpeedAlterationEffect or StunEffect.

skaldarnar committed:
feat: Use component-based effect for 'boost' effects

Extracted from #14
Based on #15

Replace the copy-paste implementation of the alteration effect giving a health boost with the common component-based base class.

:exclamation: This moves some code around and thus changes the logic. Needs careful review (although I believe that the previous implementation was erroneous).

ktksan commented:
Yep, the healthBoost alteration effect's previous implementation did not work. And simply using the new ChangeMaxHealthEvent directly where maxHealth is changed is not a good solution either because:
The Equipments module uses alteration effects in a weird way - when an equipment is unequipped then the respective alteration effect is re-applied with a zero duration. 
Which means there are some problems in an effect like this one where the change made is relatively more permanent in the sense that this effect is applied and removed only at the beginning and ending of the duration as compared to walkSpeedAlteration  which is checked for at every GetMaxSpeedEvent.
So the health boost is applied even when the item is unequipped if done that way... 
Note: these are some of the issues of the previous implementation

ktksan commented on src/main/java/org/terasology/alterationEffects/boost/HealthBoostAlterationEffect.java:
```diff
@@ -1,160 +1,77 @@
+    private void removeBoost(EntityRef entity, HealthBoostComponent boost) {
+        entity.updateComponent(HealthComponent.class, health -> {
+            // Reverse the max health boosting effect by dividing the old boost amount.
+            health.maxHealth = Math.round(health.maxHealth / (1f + 0.01f * boost.boostAmount));
```
Actually this won't change the healthBar UI, so the new maxHealth will be mapped to the original 10 icons..
I'd suggest using the new added ChangeMaxHealthEvent.. which will also take care of the health.currentHealth changes itself.
Same for all the places where maxHealth is modified in here..

skaldarnar replied:
Good point, I'll update that! :+1:

jdrueckert commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Alteration Effects](https://trello.com/c/HZ81jhY2/6-alteration-effects)

jdrueckert committed:
Merge branch 'develop' into topic/component-based-effect-boost

skaldarnar committed:
review: use ChangeMaxHealthEvent to also udpate the UI

skaldarnar commented on src/main/java/org/terasology/alterationEffects/boost/HealthBoostAlterationEffect.java:
```diff
@@ -1,160 +1,75 @@
+        component.lastUseTime = time.getGameTimeInMs();
 
         // Get the health component of this entity, and increase its max health using the health boost multiplier.
-        HealthComponent h = entity.getComponent(HealthComponent.class);
-        h.maxHealth = Math.round(h.maxHealth * (1 + 0.01f * hbot.boostAmount));
```
There is no hook to act in this position in the component-based effect. I don't see a nice way to handle this, except for doing these three steps in both `update` and `upsert`

1. revert effect
2. update component
3. apply effect

skaldarnar committed:
- `lastUseTime` is never read from HealthBoostComponent
- use ChangeMaxHealthEvent to reset health when component is removed
- reset and re-apply health boost in both update and upsert hooks
